worker/src/handlers.rs

```diff
@@ -73,4 +240,140 @@ impl PrimaryToWorker for PrimaryReceiverHandler {
+        let message = request.into_body();
+
+        let mut missing = HashSet::new();
+        let mut available = HashSet::new();
+
+        for digest in message.digests.iter() {
+            // Check if we already have the batch.
+            match self.store.read(*digest).await {
+                Ok(None) => {
+                    missing.insert(*digest);
+                    debug!("Requesting sync for batch {digest}");
+                }
+                Ok(Some(_)) => {
+                    available.insert(*digest);
+                    trace!("Digest {digest} already in store, nothing to sync");
+                    continue;
+                }
+                Err(e) => {
+                    error!("{e}");
+                    continue;
+                }
+            };
+        }
+
+        // reply back immediately for the available ones
+        if !available.is_empty() {
+            // Doing this will ensure the batch id will be populated to primary even
+            // when other processes fail to do so (ex we received a batch from a peer
+            // worker and message has been missed by primary).
+            for digest in available {
+                let message = WorkerPrimaryMessage::OthersBatch(digest, self.id);
+                let _ = self.tx_primary.send(message).await.tap_err(|err| {
+                    debug!("{err:?} {}", DagError::ShuttingDown);
+                });
+            }
+        }
+
+        if missing.is_empty() {
+            debug!(
+                "All batches are already available {:?} nothing to request from peers",
+                message.digests
+            );
+            return Ok(anemo::Response::new(()));
+        }
+
+        // Send sync request to a single node.
+        let worker_name = match self.worker_cache.load().worker(&message.target, &self.id) {
+            Ok(worker_info) => worker_info.name,
+            Err(e) => {
+                return Err(anemo::rpc::Status::internal(format!(
+                    "The primary asked us to sync with an unknown node: {e}"
+                )));
+            }
+        };
+
+        debug!(
+            "Sending WorkerBatchRequest message to {} for missing batches {:?}",
+            worker_name,
+            missing.clone()
+        );
+
+        // TODO: restore a metric tracking number of batches with inflight requests?
+        let (tx_batch_response, rx_batch_response) = oneshot::channel();
+        let message = WorkerBatchRequest {
+            digests: missing.iter().cloned().collect(),
+        };
+        self.tx_request_batches_rpc
+            .send((Some(worker_name), None, message, tx_batch_response))
+            .await
+            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?;
+        let response = rx_batch_response
+            .await
+            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?;
+        if let Ok(response) = response {
```

Comment:
Should we at least log an Err response?

Reply:
So this RPC should be expected to occasionally fail since it's sent unreliably, other nodes may be byzantine, etc. RPC errors will at least be counted in the new anemo rpc prometheus metrics. If the whole `Synchronize` fails, the primary will eventually log an error about that. Given all that do you think an additional error log here is still useful or too spammy? I am on the fence myself, so if you ask me to put it I will happily do it but at least wanted to explain the context

Reply:
I initially err on the side of logging more (trace level if you prefer), and then we can cut down if we want.

Reply:
Ok added!